src/wallet/wallet.cpp

```diff
@@ -56,16 +59,19 @@ bool RemoveWallet(CWallet* wallet)
 
 bool HasWallets()
 {
+    LOCK(cs_wallets);
     return !vpwallets.empty();
 }
 
 std::vector<CWallet*> GetWallets()
 {
+    LOCK(cs_wallets);
```

Comment:
IMO the GetWallet call with returning a vector of raw pointers is dangerous.
This is why I used the callable/lamda approach in #12587 (see https://github.com/bitcoin/bitcoin/pull/12587/files#diff-74d273024bfb20e7f09b87a23cd26f4dR41).

A concurrency mess-up with retrieving the pointers under the cs_wallet lock (copy of the vector), then continue outside of the `cs_wallet` lock may happen in the future.

Reply:
This problem obviously also is also true for `GetWallet()` and - for unloading - moving the a shared pointer and weak-unloading at a later stage is probably unavoidable.

Reply:
Right now there's no way to free wallets. Obviously we'll need to change this when we add an `unloadwallet` RPC, but for now this approach is sufficient to allow us to add a `loadwallet` and `createwallet` RPC.

Would it be sufficient to add a comment to these functions warning of the danger if we add a way to free wallets?

Reply:
Maybe a comment... I think its also okay to just be aware of this and fix it once we have a way to remove pointers from the array.

Reply:
I'm working on that with shared pointers. I can include this commit there, but IMO this is not worse than master so can go in to at least protect concurrency for loadwallet.